File: dashboard/src/lib/returnTo.ts
/**
 * safeReturnTo — same-origin return-URL guard.
 *
 * Accepts a string (or undefined) from untrusted input (query params, API
 * responses) and returns a safe, same-origin path to navigate to. Any value
 * that could redirect the user off-origin is rejected in favour of '/'.
 *
 * Accepts both a relative path (`/oauth/authorize?…`, set by the SPA's auth
 * guard) AND a same-origin absolute URL (`https://idp.example/oauth/authorize?…`,
 * which the server emits when bouncing an unauthenticated user to /login) —
 * both resolve to the current origin. The result is always a relative path so
 * the caller's navigation can never be read as protocol-relative.
 *
 * Rejects:
 *  - protocol-relative `//evil.com`;
 *  - cross-origin absolute URLs (origin ≠ current origin);
 *  - non-http(s) schemes like `javascript:` / `data:` (opaque "null" origin);
 *  - anything that resolves to a `//…` path (e.g. the `/\evil.com` trick).
 */
export function safeReturnTo(raw: string | undefined): string {
  if (!raw) return '/'

  // Protocol-relative ('//evil.com') would navigate off-origin — reject outright.
  if (raw.startsWith('//')) return '/'

  try {
    // Resolve relative paths and absolute URLs alike against the current origin.
    // Same-origin is the security boundary: cross-origin values and non-http
    // schemes (javascript:/data:, whose origin is the opaque "null") fail here.
    const resolved = new URL(raw, window.location.origin)
    if (resolved.origin !== window.location.origin) return '/'

    const path = resolved.pathname + resolved.search + resolved.hash
    // Never emit a value that could be read as protocol-relative: a "/\evil.com"
    // input can stay on-origin but resolve to a "//evil.com" path.
    if (path.startsWith('//')) return '/'
    return path
  } catch {
    return '/'
  }
}

File: pkg/server/returnto.go
package server

import (
	"net/url"
	"strings"

	"prohibitorum/pkg/configx"
)

// resolveReturnTo is the single same-origin source of truth for every
// return_to consumer. Given untrusted input and the configured issuer it
// reports the safe, same-origin RELATIVE path to navigate to and whether the
// input was usable. ok=false means the value is empty-unsafe or off-origin and
// the caller must fall back ("/" for fail-soft callers, or an error for
// fail-closed callers). It is the Go twin of the client safeReturnTo.
//
// Accepts: a same-origin absolute URL (scheme+host == issuer) OR a
// path-absolute relative ref ("/x"). Rejects: cross-origin, protocol-relative
// ("//", and the "\"-normalised "/\", "\\" tricks a browser reads as "//"),
// non-http schemes (javascript:/data:), userinfo/port origin deviations, and
// any result that resolves protocol-relative. Empty input → ("/", true).
// nil/empty-issuer config → absolute URLs cannot match and are rejected;
// relative paths still resolve (preserves the nil-config federation behaviour).
func resolveReturnTo(raw string, cfg *configx.Config) (string, bool) {
	// 1. Empty → safe default.
	if raw == "" {
		return "/", true
	}

	// 2. Reject protocol-relative including the backslash normalisation tricks:
	// //evil, /\evil, \\evil, \/evil — a browser normalises \ → / on navigation.
	if strings.HasPrefix(strings.ReplaceAll(raw, "\\", "/"), "//") {
		return "", false
	}

	// 3. Parse the input.
	u, err := url.Parse(raw)
	if err != nil {
		return "", false
	}

	// 4. Resolve the issuer origin defensively — only when config and issuer
	// are both present and parseable.
	var issuer *url.URL
	if cfg != nil && cfg.OIDC.Issuer != "" {
		if iss, ierr := url.Parse(cfg.OIDC.Issuer); ierr == nil {
			issuer = iss
		}
	}

	// 5. Absolute URL: require a known issuer AND scheme+host match.
	// This rejects javascript:/data: via scheme mismatch, cross-origin, port
	// deviations, and userinfo tricks (Go's url.Parse puts userinfo outside
	// u.Host, so "https://idp.example@evil.com/x" has u.Host == "evil.com").
	if u.IsAbs() {
		if issuer == nil {
			return "", false
		}
		if u.Scheme != issuer.Scheme || u.Host != issuer.Host {
			return "", false
		}
	} else {
		// 6. Relative ref: must be path-absolute (starts with "/") and not
		// protocol-relative (no host component).
		if u.Host != "" || !strings.HasPrefix(raw, "/") {
			return "", false
		}
	}

	// 7. Build the relative path from PARSED components (never from raw, so a
	// host cannot leak back into the result).
	path := u.EscapedPath()
	if path == "" {
		path = "/"
	}
	if u.RawQuery != "" {
		path += "?" + u.RawQuery
	}
	if u.Fragment != "" {
		path += "#" + u.EscapedFragment()
	}

	// 8. Final guard: result must be path-absolute and not protocol-relative.
	if !strings.HasPrefix(path, "/") || strings.HasPrefix(path, "//") {
		return "", false
	}

	// 9. Safe.
	return path, true
}

// validateReturnTo is the fail-soft policy wrapper used by the login ceremony
// and the consent decision: a safe same-origin relative path, or "/" for any
// empty/unsafe input. Never errors. Single source of truth shared with
// validateFederationReturnTo (which applies a fail-closed policy over the same
// resolveReturnTo core).
func validateReturnTo(raw string, cfg *configx.Config) string {
	if p, ok := resolveReturnTo(raw, cfg); ok {
		return p
	}
	return "/"
}
